### PR TITLE
Adding missing docstrings to all public functions.

### PR DESCRIPTION
**Pull request by MahmoudAlamar03**

### CODE REVIEW

File: ResSimpy/Nexus/NexusSimulator.py [hunk 1]
```diff
@@ -284,14 +295,17 @@ def new_fcs_name(self) -> Optional[str]:
 
     @property
     def write_times(self) -> bool:
+        """Sets whether the runcontrol file will expand the include files with time cards in. Defaults to True."""
```
Comment (jonnyh404) [inline ResSimpy/Nexus/NexusSimulator.py:298]: Is this really what this write_times does? If so we probably need to call it something slightly different!

File: ResSimpy/OpenGoSim/Model_Parts/OpenGoSimNetwork.py [hunk 2]
```diff
@@ -23,4 +23,5 @@ def __repr__(self) -> str:
         return "Not implemented yet"
 
     def load(self) -> None:
+        """Returns."""
```
Comment (jonnyh404) [inline ResSimpy/OpenGoSim/Model_Parts/OpenGoSimNetwork.py:26]: """Loads the OpenGoSimNetwork. Not Implemented Yet."""

File: ResSimpy/OpenGoSim/OpenGoSimSimulator.py [hunk 3]
```diff
@@ -70,10 +70,12 @@ def __init__(self, origin: str) -> None:
 
     @property
     def simulation_type(self) -> SimulationType:
+        """Returns an instance of OpenGo simulation type."""
         return self.__simulation_type
 
     @property
     def final_date(self) -> Optional[str]:
+        """Returns"""
```
Comment (jonnyh404) [inline ResSimpy/OpenGoSim/OpenGoSimSimulator.py:78]: Returns final date in the simulator.

File: ResSimpy/OpenGoSim/OpenGoSimSimulator.py [hunk 4]
```diff
@@ -70,10 +70,12 @@ def __init__(self, origin: str) -> None:
 
     @property
     def simulation_type(self) -> SimulationType:
+        """Returns an instance of OpenGo simulation type."""
```
Comment (jonnyh404) [inline ResSimpy/OpenGoSim/OpenGoSimSimulator.py:73]: OpenGoSim

File: ResSimpy/OpenGoSim/OpenGoSimSimulator.py [hunk 5]
```diff
@@ -214,11 +216,22 @@ def __load_in_well_data_block(self, remaining_text: list[str]) -> None:
         self._wells._wells.append(loaded_well)
         self._wells._wells_loaded = True
 
+    @staticmethod
```
Comment (jonnyh404) [inline ResSimpy/OpenGoSim/OpenGoSimSimulator.py:219]: Remove additional staticmethod decorator.

File: ResSimpy/OperationsMixin.py [hunk 6]
```diff
@@ -29,49 +29,89 @@ def __init__(self, parent_network: Network) -> None:
...
+            obj_to_modify(dict[str, None | str | float | int]): dictionary containing attributes to match
+            in the existing object.
+            new_properties(dict[str, None | str | float | int]): properties to switch to in the new object.
+        """
         raise NotImplementedError("Implement this in the derived class")
 
     @property
     @abstractmethod
     def table_header(self) -> str:
+        """Returns"""
```
Comment (jonnyh404) [inline ResSimpy/OperationsMixin.py:108]: Returns table header as a string

File: ResSimpy/Wells.py [hunk 7]
```diff
@@ -34,34 +40,85 @@ def _load(self) -> None:
         raise NotImplementedError("Implement this in the derived class")
 
     def get_all(self) -> Sequence[Well]:
+        """Returns a sequence of wells usually of type list."""
         raise NotImplementedError("Implement this in the derived class")
 
     def get(self, well_name: str) -> Optional[Well]:
+        """?
```
Comment (jonnyh404) [inline ResSimpy/Wells.py:47]: Gets a well by well name

File: ResSimpy/Wells.py [hunk 8]
```diff
@@ -34,34 +40,85 @@ def _load(self) -> None:
...
+
+        Args:
+            well_name (str): The well name.
+        """
         raise NotImplementedError("Implement this in the derived class")
 
     def get_df(self) -> pd.DataFrame:
+        """Returns a pandas data frame."""
```
Comment (jonnyh404) [inline ResSimpy/Wells.py:55]: Returns wells as a pandas dataframe

File: ResSimpy/Wells.py [hunk 9]
```diff
@@ -34,34 +40,85 @@ def _load(self) -> None:
...
         raise NotImplementedError("Implement this in the derived class")
 
     def modify(self, well_name: str, completion_properties_list: list[dict[str, None | float | int | str]],
                how: OperationEnum = OperationEnum.ADD) -> None:
+        """?
```
Comment (jonnyh404) [inline ResSimpy/Wells.py:60]: Modifies a completion in a named well using a list of properties. How enum determines if the completions are added, removed or modified.

File: ResSimpy/Wells.py [hunk 10]
```diff
@@ -34,34 +40,85 @@ def _load(self) -> None:
...
+
+        Args:
+            well_name (str) : The well name
+            completion_properties_list (list[dict[str, None | float | int | str]]) : ?
+            how (OperationEnum) : ?
```
Comment (jonnyh404) [inline ResSimpy/Wells.py:65]: Check the OperationEnum for valid options.

File: ResSimpy/Wells.py [hunk 11]
```diff
@@ -34,34 +40,85 @@ def _load(self) -> None:
...
+        """Returns a pandas data frame."""
         raise NotImplementedError("Implement this in the derived class")
 
     def modify(self, well_name: str, completion_properties_list: list[dict[str, None | float | int | str]],
                how: OperationEnum = OperationEnum.ADD) -> None:
+        """?
+
+        Args:
+            well_name (str) : The well name
+            completion_properties_list (list[dict[str, None | float | int | str]]) : ?
```
Comment (jonnyh404) [inline ResSimpy/Wells.py:64]: List of completions property dictionaries to pass through to each of the ADD/REMOVE/MODIFY operations.

File: ResSimpy/Wells.py [hunk 12]
```diff
@@ -34,34 +40,85 @@ def _load(self) -> None:
...
         raise NotImplementedError("Implement this in the derived class")
 
     def add_completion(self, well_name: str, completion_properties: dict[str, None | float | int | str],
                        preserve_previous_completions: bool = True, comments: Optional[str] = None) -> None:
+        """Add the new completion.
+
+        Args:
+            well_name(str): The well name.
+            completion_properties(dict[str, None | float | int | str]): Properties of the added completion.
+            preserve_previous_completions():? .
```
Comment (jonnyh404) [inline ResSimpy/Wells.py:79]: When adding new completions to a new date we can either... Simply add the completion on its own, effectively removing all previous completions or we can add it with the previous set of completions at the new timestep. 
This bool toggles that behaviour.

File: ResSimpy/Wells.py [hunk 13]
```diff
@@ -34,34 +40,85 @@ def _load(self) -> None:
...
+        """
         raise NotImplementedError("Implement this in the derived class")
 
     def remove_completion(self, well_name: str,
                           completion_properties: Optional[dict[str, None | float | int | str]] = None,
                           completion_id: Optional[UUID] = None) -> None:
+        """Removes completion.
+
+        Args:
+            well_name(str): Well name.
```
Comment (jonnyh404) [inline ResSimpy/Wells.py:93]: Well name to remove the completion from.

File: ResSimpy/Wells.py [hunk 14]
```diff
@@ -34,34 +40,85 @@ def _load(self) -> None:
...
         raise NotImplementedError("Implement this in the derived class")
 
     def modify_completion(self, well_name: str, properties_to_modify: dict[str, None | float | int | str],
                           completion_to_change: Optional[dict[str, None | float | int | str]] = None,
                           completion_id: Optional[UUID] = None,
                           comments: Optional[str] = None) -> None:
+        """Modifies the completion properties with the new values.
+
+        Args:
+            well_name (str): Well name.
```
Comment (jonnyh404) [inline ResSimpy/Wells.py:109]: Well name to modify completion from.

File: pyproject.toml [hunk 15]
```diff
@@ -69,7 +69,7 @@ ignore = ["UP007",  "ANN101", "N999", "UP035", "S105", "N802", "S106", "S107", "
     "D105", "D203", "D213", "RUF010", "I001", "B028", "COM812", "PD901",
     # To fix later:
     "C901", # Complexity
-    "D102", "D100", "D103", "D101", "D205", # Docstrings
+    "D100", "D103", "D101", "D205", # Docstrings
```
Comment (jonnyh404) [inline pyproject.toml:72]: Great work improving the codebase and enabling this rule.

**Review by jonnyh404 — CHANGES_REQUESTED** [review]
Great work improving the code base by adding all these high quality docstrings. Please have a look at my comments and add those small changes. Otherwise looks good to me!

File: ResSimpy/Completion.py [hunk 16]
```diff
@@ -122,97 +122,117 @@ def __init__(self, date: str, i: Optional[int] = None, j: Optional[int] = None,
...
         return self.__skin
 
     @property
     def depth(self) -> float | None:
+        """Returns the depth of the completion as a float."""
         return self.__depth
 
     @property
     def x(self) -> float | None:
+        """Returns the location of the well as a float."""
```
Comment (cflynn3) [inline ResSimpy/Completion.py:155]: location of the completion in the x direction. Same for others below.

**Review by jonnyh404 — APPROVED** [review]
Looks good to me. Thanks for making the changes.

**Pull request merged**